crypto/rsa/rsa_pss.c

```diff
@@ -168,19 +169,28 @@ int RSA_padding_add_PKCS1_PSS_mgf1(RSA *rsa, unsigned char *EM,
         goto err;
     }
 
-    MSBits = (BN_num_bits(rsa->n) - 1) & 0x7;
     emLen = RSA_size(rsa);
+    if (emLen <= 1)
+        goto err;
+
+    MSBits = (BN_num_bits(rsa->n) - 1) & 0x7;
     if (MSBits == 0) {
         *EM++ = 0;
         emLen--;
     }
     if (sLen == RSA_PSS_SALTLEN_MAX) {
+        if (hLen < 2 || hLen - 2 > emLen)
+            goto err;
         sLen = emLen - hLen - 2;
-    } else if (emLen < (hLen + sLen + 2)) {
+    } else if (sLen > INT_MAX - 2 - hLen ||
+               emLen < (hLen + sLen + 2)) {
         RSAerr(RSA_F_RSA_PADDING_ADD_PKCS1_PSS_MGF1,
                RSA_R_DATA_TOO_LARGE_FOR_KEY_SIZE);
         goto err;
     }
+
+    /* At this point, sLen is >= 0 */
```

Comment:
Yes, that **should** be the case, but what happens if you try this:
openssl genrsa -out rsa.pem 512
openssl dgst -sign rsa.pem -sha512 -sigopt rsa_padding_mode:pss -hex < /dev/null